5단계 - 로또(수동) PR 요청드립니다.

earth-h opened the pull request:
태경님 안녕하세요 :) 

로또 - TDD :: 5단계 - 로또(수동) PR 피드백 요청드립니다.

이번 미션에서는 4단계 이후 로또 피드백 영상을 보면서, 일부 로직 리팩토링 하면서 진행하였습니다.

최대한 TDD 방식으로 진행하려 노력하였으나, Controller나 View쪽 수정할 부분은 테스트코드가 없다보니 TDD로 구현이 어렵다라는 생각이 들었습니다..!😭

하면서 고민되었던 포인트는, Money에서 몇개의 로또가 구매가능한지 등의 계산을 하는 부분인데요
LottoCount용으로 클래스를 따로 따기는 좀 애매한거 같아서, Money에서 계속 계산을 시켰는데,
이렇게 진행해도 될지 의문이 들고 있어서 보시면서 이 부분도 피드백 부탁드립니다 :) 

감사합니다 !

earth-h committed:
refactor(domain): LottoPrize의 findLottoPrize() 메소드 리팩토링

LottoPrize
- 기존에는 0~6사이의 값이 아닌 수가 matchCount로 들어오면 NO_PRIZE로 가도록 orElse()문을 적용하였음
- 이를 orElseThrow()로 변경하여서 잘못된 값이 넘어옴 방지
  - 로또 피드백 강의 수강하면서 수정 진행한 사항

earth-h committed:
test(domain): LottoResultsTest 리팩토링

LottoResultsTest
- createLottos() 메소드 리팩토링

earth-h committed:
refactor(controller): LottoController 리팩토링

LottoController
- 파라미터가 3개인 메소드 인자 줄임

earth-h committed:
feat(domain): Money 클래스 내 메소드 추가

Money
- 주어진 돈 내에서 살 수 있는 로또 개수인지 여부 반환 메소드
  - 파라미터로 넘겨진 인자 값이 음수이면 validation
  - 파라미터로 넘겨진 인자 값이 구매할 수 있는 수량을 넘어선 경우 false 반환

earth-h committed:
refactor(domain): isBuyableLottoCount() 메소드 리팩토링

Money
- isBuyableLottoCount() 메소드의 false가 될 수 있는 부분들을 validation 처리

earth-h committed:
feat(domain): Lottos 생성자 추가

Lottos
- 수동 로또가 포함됨에 따라 이미 생성한 List<Lotto>와 추가로 generator에 의해 생성될 로또 개수를 받아 Lottos 생성

earth-h committed:
feat(domain): Money 클래스 내 메소드 추가 및 MIN_MONEY 상수 > ZERO로 변경

Money
- MIN_MONEY 대신 ZERO 변경 사유는 Money 클래스 내에서 0이 의미하는 바가 해당 돈으로 구매 가능한 count에도 쓰일 수 있어 ZERO로 변경
- maxLottoCountExclude() 메소드 생성하여 수동 로또 제외한 자동 로또 최대 구매 개수 반환

earth-h committed:
docs(domain): Lottos 생성자 docs 추가

Lottos
- Lottos의 신규로 만든 생성자 관련 docs 추가

earth-h committed:
feat(controller,view): Controller, View 로직 추가

LottoController
- 자동/수동 로또 나뉨에 따라 로직 추가

ResultView
- 자동/수동 로또 나뉨에 따라 출력 메시지 변경

InputView
- 수동 로또 생성에 따라 관련 출력 메시지 추가

earth-h committed:
refactor(controller,domain): WinningLotto 생성에 따른 리팩토링

WinningLotto
- 당첨로또(Lotto)와 보너스번호(LottoNumber)를 가지고 있는 클래스 생성
- 당첨로또와 보너스번호 모두 당첨과 관련된 값으로 한 곳에 모아 사용하기 위해 wrapping

LottoController, Lottos, LottosResults
- WinningLotto 생성에 따라 기존에 Lotto와 LottoNumber 두 파라미터 가지고 갔던 부분 리팩토링

earth-h committed:
refactor(domain): lottoPrizeMoney long형을 변경

LottoPrize
- 1등 2명만 나와도 int형을 벗어나기 때문에, long형으로 변

LottoResults
- LottoPrize 내 형변환 진행함에 따라 LottoResults도 변경

earth-h committed:
test(domain): Lottos 테스트 추가

LottosTest
- 신규로 추가된 Lottos 생성자에 대한 테스트 코드 추가

earth-h committed:
test(domain): LottosTest 리팩토링

LottosTest
- createLottos() 메소드 리팩토링

ohtaeg commented:
>  Money에서 몇개의 로또가 구매가능한지 등의 계산을 하는 부분인데요
LottoCount용으로 클래스를 따로 따기는 좀 애매한거 같아서, Money에서 계속 계산을 시켰는데,
이렇게 진행해도 될지 의문이 들고 있어서 보시면서 이 부분도 피드백 부탁드립니다 :)

개인적으로는 LottoCount를 따로 만들지 않고 Money에서 잘 수행해주셨다고 생각합니다 👍 
LottoCount를 따로 분리해서 얻는 장점이 크게 없는 것 같아요 :)
객체를 설계할 때 중요하다고 생각하는 부분중에 하나는
객체가 유지해야 하는 정보와 수행할 수 있는 행동이 외부에 군데군데 있는 것이 아닌 한 곳에 응집되어 있는지 라고 생각하는데요 :)
책임을 수행하는 데 필요한 정보를 가장 잘 알고 있는 객체에게 책임을 할당하고 그런 책임을 가지고 있는 객체들이 협력을 하는 것이 객체지향의 패러다임이지 않을까 합니다 😄

ohtaeg commented on src/main/java/lotto/domain/LottoPrize.java:
```diff
@@ -13,28 +13,35 @@ public enum LottoPrize {
-        LottoPrize lottoPrize = Arrays.stream(LottoPrize.values()).filter(prize -> prize.getMatchCount() == matchCount)
-                .findFirst()
-                .orElse(NO_PRIZE);
+        if(isNoPrize(matchCount)) {
+            return LottoPrize.NO_PRIZE;
+        }
```
isNoPrize() 메서드로 따로 분리해주신 이유가 있을까요 :) ?
5개 일치, 보너스 볼 일치를 제외하고 matchCount가 3개 미만일 경우 NO_PRIZE를 반환하는 로직도 괜찮다고 생각해서요!

earth-h replied:
안녕하세요 :) 
위와 같이 로직 변경을 진행한 사유는, 4단계 이후 로또 피드백 영상을 보면서 validation이 필요한 부분이 있다고 생각하였기 때문입니다! 

결론부터 말씀드리자면, orElseThrow() 구문 추가로 인해, matchCount가 1이거나 2일 때 에러를 throw하게 되어 그 부분을 막고자 앞단에 조건을 하나 더 걸었다고 봐주시면 됩니다 ! 좀 과할까..?싶기는 했지만, 최대한 막아보자라는 마음으로 진행하긴 했습니다 ㅎㅎ..🙄

```java

    public static LottoPrize findLottoPrize(int matchCount, boolean isMatchBonusLottoNumber) {
        if(isNoPrize(matchCount)) {
            return LottoPrize.NO_PRIZE;
        }

        if(isSecondPrize(matchCount, isMatchBonusLottoNumber)) {
            return LottoPrize.SECOND;
        }

        return Arrays.stream(LottoPrize.values()).filter(prize -> prize.getMatchCount() == matchCount)
                .findFirst()
                .orElseThrow(IllegalArgumentException::new);
    }
```

- `findLottoPrize()` 메소드가 public이기 때문에, `findLottoPrize()` 메소드의 파라미터를 직접 넣어 호출하게 되면, matchCount에는 int이면 어떤 값이던 들어갈 수 있게 됩니다. 
  - (로직 흐름상 0~6사이값만 들어가야 정상이나, 누군가 직접 저 메소드를 사용하게 된다면, 잘못된 숫자가 들어갔을 때 NO_PRIZE가 나올 것이라, validation을 해야하지 않을까 했습니다 ㅎㅎ)
  - 기존처럼 `return Arrays.stream(~).orElse(NO_PRIZE)`를 하게 되면, 0보다 작거나 6보다 큰 수가 들어갈 수 있어 `orElseThrow()`를 진행했습니다!
  - 이렇게 `orElseThrow()`를 하는 과정에서, stream() 로직 내에 있는 `prize.getMatchCount() == matchCount`에 의해 matchCount가 1개거나, 2개일때에는 orElseThrow()로 가게 되어서 isNoPrize() 여부를 앞에서 확인해주게 되었습니다 ! 

관련된 테스트코드를 LottosPrizeTest.java에 작성해둔 부분이 있습니다 :) 참고 부탁드립니다 ㅎㅎ
```java
    @Test
    void 로또_번호가_0개에서_2개_동일하면_NO_PRIZE() {
        assertAll(
                () -> assertThat(LottoPrize.findLottoPrize(2, true)).isEqualTo(LottoPrize.NO_PRIZE),
                () -> assertThat(LottoPrize.findLottoPrize(1, true)).isEqualTo(LottoPrize.NO_PRIZE),
                () -> assertThat(LottoPrize.findLottoPrize(0, true)).isEqualTo(LottoPrize.NO_PRIZE)
        );
    }

    @Test
    void 로또_번호가_6개_초과로_동일하면_에러_발생() {
        assertThatThrownBy(() -> LottoPrize.findLottoPrize(7, true)).isInstanceOf(IllegalArgumentException.class);
        assertThatThrownBy(() -> LottoPrize.findLottoPrize(7, false)).isInstanceOf(IllegalArgumentException.class);
    }

    @Test
    void 로또_번호가_음수_개수만큼_동일하면_에러_발생() {
        assertThatThrownBy(() -> LottoPrize.findLottoPrize(-1, true)).isInstanceOf(IllegalArgumentException.class);
        assertThatThrownBy(() -> LottoPrize.findLottoPrize(-1, false)).isInstanceOf(IllegalArgumentException.class);
    }
```

ohtaeg replied:
> 잘못된 숫자가 들어갔을 때 NO_PRIZE가 나올 것이라, validation을 해야하지 않을까 했습니다 ㅎㅎ

아하 덕분에 이해가 되었습니다! 꼼꼼하게 세세한 부분까지 고려를 잘해주셨네요 👍

ohtaeg commented on src/main/java/lotto/domain/Lottos.java:
```diff
@@ -23,6 +23,20 @@ public Lottos(Money money, LottoNumberGenerator lottoNumberGenerator) {
+     * @param lottoNumberGenerator
+     * @param manualLottos 수동으로 기생성된 로또
+     */
+    public Lottos(int maxLottoCount, LottoNumberGenerator lottoNumberGenerator, List<Lotto> manualLottos) {
```
Lottos 객체를 만들어 주신만큼 List\<Lotto\>로 다루지 말고 Lottos로 다루면 좋을 것 같아요 :)

ohtaeg commented on src/main/java/lotto/domain/Money.java:
```diff
@@ -53,6 +53,16 @@ public int maxLottoCount() {
         return (int) money / LOTTO_PRICE;
     }
 
+    public boolean isBuyableLottoCount(int lottoCount) {
```
>  Money에서 몇개의 로또가 구매가능한지 등의 계산을 하는 부분인데요
LottoCount용으로 클래스를 따로 따기는 좀 애매한거 같아서, Money에서 계속 계산을 시켰는데,
이렇게 진행해도 될지 의문이 들고 있어서 보시면서 이 부분도 피드백 부탁드립니다 :)

개인적으로는 LottoCount를 따로 만들지 않고 Money에서 잘 수행해주셨다고 생각합니다 👍 
LottoCount를 따로 분리해서 얻는 장점이 크게 없는 것 같아요 :)
LottoCount가 어떤 목적을 위해 생성되었는지, **어떤 일을 하기 위해 생성되었는지** 알기 힘들 수 있을 것 같아요 :)
게다가 객체가 유지해야 하는 정보와 수행할 수 있는 행동이 외부에 군데군데 있는 것이 아닌 한 곳에 응집되어 있는 것도 중요하다고 생각하구요!
현실세계의 돈과 소프트웨어 세계에서의 돈은 다르기 때문에 Money에서 충분히 count 관련해서 로직을 처리할 수 있다고 생각합니다 ㅎㅎ

ohtaeg commented on src/main/java/lotto/domain/WinningLotto.java:
```diff
@@ -0,0 +1,29 @@
+
+import common.constant.ErrorCode;
+
+public class WinningLotto {
```
4단계 피드백을 참고하여 리팩토링 👍 
이전 설계도 좋고 지금도 좋다고 생각합니다 :)

ohtaeg commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -42,34 +49,81 @@ private Money getMoney() {
-        printPurchasingLottoCount(money.maxLottoCount());
-        Lottos lottos = new Lottos(money, new RandomLottoNumberGenerator());
+        int manualLottoCount = getManualLottoCount(money);
+        List<Lotto> manualLottos = createManualLottos(manualLottoCount);
```
같은 피드백 입니다 :)
https://github.com/next-step/java-lotto-pro/pull/842#discussion_r1009617659

ohtaeg commented on src/main/java/lotto/domain/Lottos.java:
```diff
@@ -31,11 +45,11 @@ public Money findTotalPrice() {
         for(Lotto lotto: lottos) {
-            int matchCount = lotto.findLottoMatchCount(winningLotto);
-            boolean isMatchBonusLottoNumber = lotto.isMatchLottoNumber(bonusLottoNumber);
+            int matchCount = winningLotto.findLottoMatchCount(lotto);
+            boolean isMatchBonusLottoNumber = winningLotto.isMatchBonusLottoNumber(lotto);
             LottoPrize lottoPrize = LottoPrize.findLottoPrize(matchCount, isMatchBonusLottoNumber);
```
이 부분은 한번 고민해보면 좋을 것 같아요 :)

- findLottoMatchCount()
- isMatchBonusLottoNumber()

이 두개의 메서드를 통해 LottoPrize에 바로 접근을 하기 때문에 굳이 Lottos를 거쳐 LottoPrize를 호출하는 것 보다
당첨 번호들을 관리하면서 당첨을 판단하는 책임을 가진 WinningLotto가 
당첨 판단을 하여 결과 값으로 LottoPrize를 반환하는 것은 어떨까 싶은데 지수님은 어떻게 생각하시나요 😄 ?

earth-h replied:
이 부분은 저도 고민을 했던 포인트인데요!

제가 WinningLotto 클래스를 만들면서, WinningLotto를 만드는게 맞을까?라는 생각을 했었습니다.. 
"당첨로또"에 보너스볼도 함께 포함되어야 하는건 맞을 것 같은데, matchCount를 계산하거나, match여부를 확인하는 건 WinningLotto만의 기능일까? 이런 생각이 계속 들더라구요..! 그래서, 기존 로직인 Lotto의 findLottoMatchCount()와 isMatchLottoNumber()를 그대로 사용하고자 했습니다.

그런데, 태경님 말씀을 들어보니, 제가 WinningLotto라는 클래스의 책임을 좀 작게 판단을 했던 것 같습니다..! 저는 WinningLotto를 당첨번호를 관리하는 객체라고만 생각을 해버렸네요 ㅎㅎ 
사실상 기존에 Lotto에 있던, 하기 메소드들에서 Lotto라는 클래스 인자가 필요한 부분은 없기 때문에, 말씀해주신것처럼 WinningLotto에 메시지를 보내서 WinningLotto 내에서 답변을 보내는 것이 더 정확한 책임 분리라는 생각이 듭니다! 

[ Lotto 책임 ]
- 로또 관리
  - "로또"라는 객체에 걸맞는 validation

[ WinningLotto 책임 ]
- 당첨 로또 관리
- 당첨 판단

피드백 감사합니다 :) 

==================================================================================
제가 태경님 피드백을 잘못 이해를 했던것 같습니다 ㅎㅎ..!
WinningLotto를 수정해보면서, WinningLotto를 수정해도 Lotto에 의존이 됨을 확인하고 이상하다 생각이 들어서..보니,
Lottos를 중간에 두고 WinningLotto에 메시지를 보내서 얻은 값들을 기반으로 LottoResults를 생성함을 깨달았습니다!
피드백 참고하여서 불필요한 객체를 통하지 않도록 로직 변경해보겠습니다 :) 

==================================================================================
기존에는 Lottos > WinningLotto > Lotto를 호출하여서 결과를 도출했는데, 
Lottos에서는 WinningLotto > Lotto에 의해 생성된 결과만 조립하는 역할을 하니, 해당 조립을 WinningLotto에서 하면
Lottos는 굳이 관여하지 않아도 되네요..! 

피드백 감사합니다 !

ohtaeg replied:
제가 피드백을 잘못드렸나? 생각하며 조마조마 하면서 봤네요 ㅎㅎ
제 설명이 부족했음에도 불구하고 잘 이해해주시고 반영해주셔서 감사드립니다 😄  🙇

ohtaeg commented on src/main/java/lotto/domain/Lottos.java:
```diff
@@ -31,11 +45,11 @@ public Money findTotalPrice() {
     }
 
-    public LottoResults createLottoResults(Lotto winningLotto, LottoNumber bonusLottoNumber) {
+    public LottoResults createLottoResults(WinningLotto winningLotto) {
```
WinningLotto가 이미 Lotto의 존재를 알고 있는 상태에서
Lotto가 WinningLotto도 알게된 양방향 의존 관계로 되어있는데요 :)
양방향 관계를 단방향으로 개선해볼 수 있을까요~?

ohtaeg reviewed (changes requested):
5단계 미션도 구현을 잘해주셨어요! 👍
이전 미션들을 잘 진행해주셔서 큰 구조 변경없이 구현을 완료해주셨네요 👍
마지막 미션인만큼 마지막으로 고민해볼 포인트가 존재하여 Request changes 하였어요 :)
고민해보신 후 반영해주실 부분이 있다면 반영해주시고 재요청 주시면 감사하겠습니다!
마지막까지 화이팅입니다 🔥

ohtaeg commented on src/main/java/lotto/domain/Lottos.java:
```diff
@@ -31,11 +45,11 @@ public Money findTotalPrice() {
 
-    public LottoResults createLottoResults(Lotto winningLotto, LottoNumber bonusLottoNumber) {
+    public LottoResults createLottoResults(WinningLotto winningLotto) {
         Map<LottoPrize, Integer> lottoResults = new HashMap<>();
```
Lottos가 LottoResult의 내부 구조인 Map<LottoPrize, Integer>를 자세하게 알 필요가 있을지 고민해보시면 좋을 것 같아요 😄

earth-h replied:
LottoResults는 사실상 WinningLotto > Lotto 사이에서 얻어지는 값들로 만들어지는 결과이기 때문에
Lottos는 해당 구조를 몰라도 되겠네요..! 

감사합니다 👍🏻

earth-h committed:
refactor(domain,controller): List<Lotto> 대신 Lottos 가지고 있을 수 있게 리팩토링

LottoController
- 수동로또 생성 시, List<Lotto>를 반환하는 대신, List<Lotto>를 가지고 있는 일급 콜렉션 객체인 Lottos 반환하도록 함

Lottos
- 기존에 List<Lotto>를 인자로 받던 생성자를 제거하고, mergeLottos()라는 이름을 가진 정적 팩토리 메소드 생성하여 좀 더 직관적으로 Lottos 생성함
  - List<Lotto> 인자 대신 Lottos 받도록 함

LottosTest
- 신규로 생성된 정적 팩토리 메소드 사용하도록 테스트 리팩토링

earth-h committed:
refactor(domain): Lottos 내 Money를 인자로 받던 생성자 제거

Lottos
- 수동 로또 로직 생성으로 인해 테스트로직에서만 사용되던 Lottos 생성자 제거

LottosTest, MoneyTest
- Lottos 생성자 제거에 따른 테스트 로직 리팩토링

earth-h committed:
refactor(domain,controller): Lottos의 createLottoResults() 메소드 WinningLotto로 위치 변경

Lottos
- Lottos의 createLottoResults()는 사실상 WinningLotto와 Lotto간 메시지 보냄을 통해 얻어진 결과값을 다시 LottoResults에 보내어 조립을 요청하는 컨트롤타워(?) 역할
- 해당 컨트롤타워 역할은 WinningLotto에서 하는게 좀 더 정확한 역할 분리
  - WinningLotto란, 당첨로또를 관리하고 당첨여부를 체크해주는 특별한 Lotto

WinningLotto
- Lottos에 존재하던 createLottoResults()를 WinningLotto로 리팩토링
  - 사실상, Lottos에 있을 때와 로직은 다를 바 없음. 즉, Lottos에 있을 이유가 없었던 것..

LottoController
- createLottoResults() 메소드가 속한 클래스 변경에 따른 해당 메소드 주체와 파라미터 변경

LottosTest, LottoResultsTest
- createLottoResults() 메소드가 속한 클래스 변경에 따른 테스트 로직 리팩토링

earth-h commented:
태경님 안녕하세요 :) 

먼저, 늦은 시간에도 피드백 꼼꼼히 남겨주신 점 정말정말 감사합니다 ! 🥰

주신 피드백 바탕으로 로직 리팩토링 진행하였습니다 ! 
- List<Lotto> 파라미터 받는 부분 Lottos로 변경
  - 사실, 편하려고 List<Lotto> 눈감고 썼던 부분이었던지라 ㅎㅎ.. 딱 걸렸네요..! 
  - List<Lotto>일 때는 Lottos 생성할 때 일반 생성자 사용해도 괜찮을 것 같았는데, Lottos를 생성할 때 인자로 Lottos가 넘어가면 뭔가.. 정확한 이름을 가진 메소드를 통해 Lottos를 생성해야할 것 같다는 생각이 들어 mergeLottos라는 정적 팩토리 메소드를 만들어보았습니다 :) 

- Lottos 내 createLottoResults() 메소드 WinningLotto로 위치 변경
  - 말씀해주신 것처럼, Lottos에서는 LottoResults가 어떻게 생겼는지 알 이유가 없었습니다..! 기존에 WinningLotto가 없었을때에는 Lottos에서 조립해서 보내는게 맞지만,, WinningLotto가 생긴 지금 Lottos의 역할 중 당첨과 관련된 역할은 가져감이 맞는 것 같습니다 :) 
  -  실제로 WinningLotto로 createLottoResults() 메소드를 옮겨보니, 변경할 부분이 없더라구요! 그만큼 Lottos는 그냥 중간다리역할 뿐이었던 것이죠🤣

> 현실세계의 돈과 소프트웨어 세계에서의 돈은 다르기 때문에 Money에서 충분히 count 관련해서 로직을 처리할 수 있다고 생각합니다 ㅎㅎ

그리고 제가 궁금해했던 부분에 대해 피드백 주신 부분 중 위 내용을 본 순간 '아, 현실세계 돈이라고 생각해버렸구나!' 라는 생각이 들었습니다 ㅎㅎ 일깨워주셔서 감사합니다 :) 

한 가지 더 궁금한 점이 있어 여쭤보고자 합니다 ! 
- A 클래스와 B 클래스가 연관되어있다!를 판단하실 때 로직을 본 순간 딱 보이시는지 궁금합니다 ㅎㅎ.. 그림을 그려봐야하는 것인지.. 어떻게 해야 좀 더 빠르게(?) 의존관계를 파악할 수 있을까 해서요 :) 아니면, 의존관게를 파악해야겠다라는 생각으로 접근을 애초부터 해야할지..?! 이런 부분에 대해 조언 부탁드립니다 :)

earth-h committed:
refactor(test): 마지막 newline 추가

POSIX 표준에 따른 마지막 줄 생성

ohtaeg commented on src/main/java/lotto/domain/Lottos.java:
```diff
@@ -1,47 +1,37 @@
+     * @param lottoNumberGenerator
+     * @param manualLottos 수동으로 기생성된 로또
+     */
+    public static Lottos mergeLottos(int maxLottoCount, LottoNumberGenerator lottoNumberGenerator, Lottos manualLottos) {
```
> List 파라미터 받는 부분 Lottos로 변경
>> 사실, 편하려고 List 눈감고 썼던 부분이었던지라 ㅎㅎ.. 딱 걸렸네요..!
List일 때는 Lottos 생성할 때 일반 생성자 사용해도 괜찮을 것 같았는데, Lottos를 생성할 때 인자로 Lottos가 넘어가면 뭔가.. 정확한 이름을 가진 메소드를 통해 Lottos를 생성해야할 것 같다는 생각이 들어 mergeLottos라는 정적 팩토리 메소드를 만들어보았습니다 :)

매의 눈으로 봤습니다ㅎㅎ  👀
mergeLottos라는 정적 팩토리 메소드를 잘 추출해주셨네요 👍 
저도 지수님과 같이 _Lottos를 생성할 때 인자로 Lottos가 넘어가면 뭔가.. 정확한 이름을 가진 메소드를 통해 Lottos를 생성해야할 것 같다_ 라는 생각을 똑같이 했었는데요! 이 부분에 대해 우선 생성자 타입을 Lottos로 변경해주시면 메서드를 분리하도록 추가 피드백을 드릴려고 했었는데 잘해주셨네요 👍 👍

ohtaeg commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -42,34 +50,81 @@ private Money getMoney() {
+    private Lottos generateManualLottos(int manualLottoCount) {
+        List<Lotto> lottos = new ArrayList<>();
+        for(int count = 0; count < manualLottoCount; count++) {
+            lottos.add(Lotto.generateLotto(new ReadLineLottoNumberGenerator(readLine())));
```
LottoNumberGenerator에서는 다음과 같은 추상 메서드가 선언되어 있는데요 :)
`List<LottoNumber> generateLottoNumbers();`

반환 타입이 Lotto가 아닌 List<LottoNumber>로 해주신 의도가 있으실까요~?
다른 의도가 없으시다면 해당 피드백과 같은 의도로 추후에 혹은 다른 미션에서 Wrapping한 객체로 
관리하도록 해주시면 좋을 것 같습니다 😄 

https://github.com/next-step/java-lotto-pro/pull/842#discussion_r1009617659

earth-h replied:
LottoNumberGenerator에서 List<LottoNumber>로 관리하던 이유는, **NumberGenerator에게 "숫자생성기"라는 역할만 부여**해주기 위함이었습니다😀

generateLottoNumbers()에 의해 List<LottoNumber>를 생성한 후, Lotto 생성자 내에서 validation을 통해 lotto에 걸맞는 숫자인지(개수는 6개인지/중복숫자는없는지)를 판단하고자 했습니다 :) 

```java
    private Lotto(List<LottoNumber> lottoNumbers) {
        validateLotto(lottoNumbers);
        this.lottoNumbers = lottoNumbers;
    }

    public static Lotto generateLotto(LottoNumberGenerator lottoNumberGenerator) {
        List<LottoNumber> lottoNumbers = lottoNumberGenerator.generateLottoNumbers();
        return new Lotto(lottoNumbers);
    }
```

말씀해주신것처럼, 아래와 같이 generateLottoNumbers()메소드를 바꿀 수도 있겠지만, 그렇게 되면 저 메소드 하나에 역할이 2개가 되는 것이 아닌가 해서요! 혹시 제가 잘못 이해했다면 말씀 부탁드립니다 ㅎㅎ 

```java
    @Override
    public Lotto generateLottoNumbers() {
        // LottoNumbers 생성
        Collections.shuffle(lottoNumbers);
        List<LottoNumber> randomLottoNumbers = new ArrayList<>();
        for(int i = 0; i < LOTTO_NUMBER_COUNT; i++) {
            randomLottoNumbers.add(lottoNumbers.get(i));
        }
        Collections.sort(randomLottoNumbers);
        // Lotto 생성
        return new Lotto(randomLottoNumbers);
    }
```

ohtaeg replied:
아하 비슷한 의견인 것 같아요 👍 
저는 Lottos를 생성하는 행위 자체가 validation 및 로또를 생성해서 정렬된 데이터까지 가지는 것을 하나로 봤었거든요 :)
현재 구조나 말씀해주신 generateLottoNumbers()로 해주셔도 큰 차이가 없네요 👍 
다른 의도가 있었는지 드렸던 질문이였습니다 ㅎㅎ
답변 감사드립니다!!

ohtaeg commented on src/main/java/lotto/domain/WinningLotto.java:
```diff
@@ -0,0 +1,45 @@
+        return lotto.isMatchLottoNumber(bonusLottoNumber);
+    }
+
+    public LottoResults createLottoResults(Lottos lottos) {
```
> Lottos 내 createLottoResults() 메소드 WinningLotto로 위치 변경


피드백 반영 💯 
이 부분에 대해 드렸던 피드백은 리뷰어분들마다 다르게 생각할 수도 있다고 생각이 됩니다!
사람마다 생각이 다른 부분이라 정답이 없는 부분인 것 같아요 :)

ohtaeg commented on src/test/java/lotto/domain/WinningLottoTest.java:
```diff
@@ -0,0 +1,33 @@
+import common.constant.ErrorCode;
+import org.junit.jupiter.api.Test;
+
+public class WinningLottoTest {
```
👍 👍

ohtaeg commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -1,7 +1,10 @@
 package lotto.controller;
 
```
> A 클래스와 B 클래스가 연관되어있다!를 판단하실 때 로직을 본 순간 딱 보이시는지 궁금합니다 ㅎㅎ.. 그림을 그려봐야하는 것인지.. 어떻게 해야 좀 더 빠르게(?) 의존관계를 파악할 수 있을까 해서요 :) 아니면, 의존관게를 파악해야겠다라는 생각으로 접근을 애초부터 해야할지..?! 이런 부분에 대해 조언 부탁드립니다 :)

우선, 설계하면서 개발하신 분의 시점과 설계를 파악하기 위해 분석하는 시점은 다를 수 있다고 생각해요 :)
코드를 직접 작성하시다보니 의존성을 놓칠수 밖에 없다고 생각합니다 😄 
잘해주셔서 크게 드릴 조언이 없네요 ㅎㅎ; 

개인적으로는 큰 그림의 설계를 먼저 하고 개발하거나 분석하는 편이에요!
간단한 설계는 아이패드에 혹은 draw.io에 uml을 그려가면서 하고
실제 개발할때는 plantUML을 이용해서 클래스 다이어그램, 액티비티 다이어그램, 시퀀스 다이어그램은 만드는 편이에요!

설계 및 분석할때는 세세하게 그리지 않고 먼저 크게크게 그려가면서 틀을 잡는 편이에요!
그리면서 고민하는 부분은 누가 이 객체를 생성하는게 좋을까? 이 객체가 생성되면 언제 사라질까? 등의 
객체 라이프 사이클을 생각하면서 관계(연관, 의존)를 고려하는 것 같아요 :)

주셨던 질문에 답변이 되었으면 좋겠네요 🙏

earth-h replied:
업무할 때에 보통 시간에 쫓기면서 기능을 하나하나 만드는 느낌이라, 말씀해주셨던 고려 포인트들을 잊고 지냈던 것 같습니다!

> 누가 이 객체를 생성하는게 좋을까? 이 객체가 생성되면 언제 사라질까? 등의
객체 라이프 사이클을 생각하면서 관계(연관, 의존)를 고려

말씀해주셨던 포인트들 기억하면서 저도 업무할 때에 큰 그림의 설계를 그리면서 진행해야겠다는 생각이 듭니다 ! 마지막까지 피드백 감사합니다 ㅎㅎ 이번 미션 통해서 로직 구현도 구현이지만, 요구사항을 받았을 때 대하는 자세 등 또한 배울 수 있는 시간이었습니다. 

감사합니다 ! 👍🏻

ohtaeg replied:
저 또한 지수님을 통해 많은 생각을 하게된 좋은 계기가되어 감사하다는 말씀을 전달드립니다 :)
남은 미션도 화이팅입니다 🔥

ohtaeg reviewed (approved):
로또 미션 완주를 축하드려요!! 🎉
마지막까지 피드백 반영을 잘해주셨습니다 👍 

제가 드린 리뷰들이 정답이 아니지만 지수님께서 앞으로 개발하시면서
승승장구하는데에 조금이나마 보탬이 되었으면 합니다.
고생 많이하셨습니다! 남은 미션들 화이팅이며 함께해서 영광이였습니다 🙇
